### PR TITLE
Update matplotlib to 3.5 in conda

### LINKED ISSUE #33529: Update matplotlib to 3.5.* in conda
Currently known issues with matplotlib 3.5 in workbench:
- [x] Right click on plot raises an exception
- [x] Save plots fails semi-silently (load probably also fails)
- [x] set_windowtitle is deprecated and will be removed in 3.6
- [x] Interfaces -> Direct -> DGSPlanner doesn't open
- [x] Warning when opening DGSPlanner
- [x] Right click workspace -> Plot -> 3D -> Surface Exception
- [x] 3D -> Wireframe
- [x] 3D -> Contour
- [x] Warning when creating 3D plot on workspace
- [x] On close with a plot open an error occurs
- [x] Check system test for opening interfaces fails appropriately
- [x] Ensure plots are opened on main thread.
    - [x] Load plots

Perform some detailed tests to make sure it works.

### PR DESCRIPTION
**Pull request by Pasarus**
**Description of work.**
A lot of small patches to ensure matplotlib 3.5.* compatibility with workbench

**To test:**
- Using your mantid conda developer environment, build, and launch workbench. Ensure you have matplotlib 3.5.1 or higher.
- Load a workspace capable of 3D plots
- Open two, 3D plots surface and wireframe
- Save your project
- Load project
- 2 new plots should have opened the same as the two originals
- No warnings, errors, or exceptions should have occured.

Fixes #33529 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/LoadEXED.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def PyExec(self):
         # nbins=int(parms_dict['NTC'])
         xdata = np.array(det_tbc)
         xdata_mon = np.linspace(xdata[0], xdata[-1], len(xdata))
-        ydata = data.astype(np.float)
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/LoadEXED.py:88]: Can I ask about these `numpy` changes? They don't seem related to `matplotlib`?

File: mantid-developer-linux.yml [hunk 2]
```diff
@@ -39,6 +40,7 @@ dependencies:
   - scipy>=1.6.2
   - setuptools=48.0.0 # Pinned purposefully due to incompatibility with later versions
   - sphinx<=3.5.4 # Pinned at 3.5.4 and below due to styling changes in 4.0, and 4.3 has a bug fix which imports a lot of matplotlib code (probably correctly, based on what we have in the api/python/mantid/plots/index.html file) and fails the doctests.
+  - sphinx_bootstrap_theme<=0.7.1
```
Comment (martyngigg) [inline mantid-developer-linux.yml:43]: I notice in pip we had `>=0.7.1` but here we have `<=`. Was that intentional?

**Review by martyngigg — COMMENTED** [review]
Just a couple of initial questions...

File: Framework/PythonInterface/mantid/plots/mantidaxes.py [hunk 3]
```diff
@@ -1292,25 +1292,21 @@ class MantidAxes3D(Axes3D):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
-        # Remove the connection for when you click on the plot as this is dealt with in figureinteraction.py to stop
-        # it interfering with double-clicking on the axes.
-        self.figure.canvas.mpl_disconnect(self._cids[1])
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/PythonInterface/mantid/plots/mantidaxes.py:1297]: We did

File: Framework/PythonInterface/mantid/plots/plotfunctions.py [hunk 4]
```diff
@@ -202,10 +202,10 @@ def plot(workspaces, spectrum_nums=None, wksp_indices=None, errors=False,
     if waterfall:
         ax.set_waterfall(True)
 
-    if not overplot:
-        fig.canvas.set_window_title(figure_title(workspaces, fig.number))
+    if not overplot and fig.canvas.manager is not None:
```
[the remark this answers is not in the document]
Comment (Pasarus) [reply Framework/PythonInterface/mantid/plots/plotfunctions.py:205]: Apparently, it was always possible, I was reading how some other projects had handled this migration and I came across this recommendation. Not sure if it's the case in mantid, but it is far better, to be safe than sorry later.

File: Framework/PythonInterface/plugins/algorithms/LoadEXED.py (hunk 1 above)
Comment (Pasarus) [reply Framework/PythonInterface/plugins/algorithms/LoadEXED.py:88]: They popped up whilst working on matplotlib, and were throwing warnings, technically not related, but still an issue in our conda packages.

File: mantid-developer-linux.yml (hunk 2 above)
Comment (Pasarus) [reply mantid-developer-linux.yml:43]: I changed it intentionally to match how we are handling sphinx. We probably don't want the theme to change underneath us.

File: Framework/PythonInterface/plugins/algorithms/LoadEXED.py (hunk 1 above)
Comment (martyngigg) [reply Framework/PythonInterface/plugins/algorithms/LoadEXED.py:88]: Ok. Do we know if there are tests against the current versions of `numpy` such that these will still work?

Comment (Pasarus) [reply Framework/PythonInterface/plugins/algorithms/LoadEXED.py:88]: According to the advice from the warning they mean the same thing. np.float == float.

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py [hunk 5]
```diff
@@ -206,12 +208,12 @@ def get_axes_object_variable(ax):
...
-    try:
-        if ax.numRows > 1:
-            ax_object_var += "[{row_num}]".format(row_num=row_num(ax))
-        if ax.numCols > 1:
-            ax_object_var += "[{col_num}]".format(col_num=col_num(ax))
-    except AttributeError:
-        # No numRows or NumCols members, so no list use the default
-        pass
+
+    if (LooseVersion('3.1.3') < LooseVersion(matplotlib.__version__) and ax.get_gridspec().nrows > 1) or \
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py:212]: This feels a little messy. Is there a reason we can't just use the `hasattr` without needing to check the version?

```
if hasattr(ax, "get_gridspec"):
    num_rows = ax.get_gridspec().nrows
    num_cols = ax.get_gridspec().ncols
else:
    num_rows = ax.numRows
    num_cols = ax.numCols
...
```    
```

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py [hunk 6]
```diff
@@ -66,24 +64,42 @@
                           "ADS.retrieve(...)\n"
                           "\n"
                           "fig, axes = plt.subplots(...)\n"
-                          "axes[0].plot(...)\n"
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py:69]: Can I ask if we are now missing tests for generating subplots with many Axes?

File: qt/python/mantidqt/mantidqt/plotting/functions.py [hunk 7]
```diff
@@ -342,7 +345,11 @@ def plot_surface(workspaces, fig=None):
             fig.clf()
             ax = fig.add_subplot(111, projection='mantid3d')
         else:
-            fig, ax = plt.subplots(subplot_kw={'projection': 'mantid3d'})
+            if LooseVersion('3.4.0') <= LooseVersion(matplotlib.__version__):
+                fig, ax = plt.subplots(subplot_kw={'projection': 'mantid3d', "auto_add_to_figure": False})
```
Comment (martyngigg) [inline qt/python/mantidqt/mantidqt/plotting/functions.py:349]: On the surface this seems strange to say `auto_add_to_figure` is `False` and then add it? Could you help me understand why? If necessary could you add a comment?

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py (hunk 5 above)
Comment (Pasarus) [reply qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py:212]: Because it prints a deprecation warning, whilst it still has the attribute if that makes sense?

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py (hunk 6 above)
Comment (Pasarus) [reply qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py:69]: I added one that explicitly tests that, it's in this file. I didn't like how it implicitly tested tiled plots instead of explicitly so I added the new test.

File: qt/python/mantidqt/mantidqt/plotting/functions.py (hunk 7 above)
Comment (Pasarus) [reply qt/python/mantidqt/mantidqt/plotting/functions.py:349]: As of 3.4.0 the behaviour of automatically adding to figure for Axis3D has been deprecated and prints a warning explaining such, then suggesting to do this. This way the behaviour is maintained that we had prior, whilst hiding the deprecation warning, and in future (matplotlib 3.6+) we will probably need to remove `auto_add_to_figure`.

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py (hunk 5 above)
Comment (martyngigg) [reply qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py:212]: If `get_gridspec` existed before the deprecation was added (I think it was? (https://github.com/matplotlib/matplotlib/commit/18d20e556ef94105869fe874ea52607c34c08b8d)), won't it always hit that first and not the `numRows` one there?

Comment (Pasarus) [reply qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py]: Just tried it out and it seems fine from what I can tell. Let me know if the version fixes are ok.

Comment (Pasarus) [reply qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py]: The CI has rediscovered the issue! 3.1.3 does have gridspec, but not nrows, and ncols.... I mean we can probably just do `hasattr(ax.get_gridspec(), nrows)` and `hasattr(ax.get_gridspec(), ncols)`.

File: docs/source/release/v6.4.0/mantidworkbench.rst [hunk 8]
```diff
@@ -13,5 +13,11 @@ Bugfixes
 --------
 - In DrILL interface, scrolling down in the settings dialog no longer affects the comboxes
 - Cleaned up the appearance of the main window and the Indirect, Reflectometry, and Engineering GUIs on macOS.
+- Fixed where matplotlib 3.5 caused 3D plots to throw a warning in multiple places.
```
Comment (martyngigg) [inline docs/source/release/v6.4.0/mantidworkbench.rst:16]: Can we move these notes out to a separate file as per the annoucements in slack?

**Review by martyngigg — CHANGES_REQUESTED** [review]
This looks good. We just need to move out the release notes to the new scheme.

File: docs/source/release/v6.4.0/mantidworkbench.rst (hunk 8 above)
Comment (Pasarus) [reply docs/source/release/v6.4.0/mantidworkbench.rst:16]: My understanding is it should be fine as long as it is merged before Friday?

**Review by martyngigg — APPROVED** [review]

**Comment by sf1919** [comment]
@mantidproject/gatekeepers please can we get this looked at today? Today is the deadline for merging in release notes the old way otherwise it will hold it up the release note project. I also note this is blocking #33561
